apps/dashr-web-trader/utils/helper-functions.R

```diff
@@ -0,0 +1,1235 @@
+  )
+}
+
+# Generate the empty news table (content filled by updateNews function)
+generateNewsTable <- function(dataframe, max_rows = 10){
+  n <- min(max_rows, nrow(dataframe))
+  rows <- lapply(seq(1, n), function(i){
+    htmlTr(
+      htmlTd(
+        htmlA(
+          dataframe[i, "title"],
+          href = dataframe[i, "url"],
+          style = list(
+            textDecoration = "none"
+          )
+        ),
+        className = "newsRow",
+        style = list(
+          fontSize = 15,
+          padding = "0px 10px 0px 10px"
+        ),
+      )
+    )
+  })
+  htmlDiv(
+    list(
+      htmlP(
+        list(
+          "Headlines",
+          htmlBr(),
+          htmlSpan(
+            id = "news_update",
+            children = paste(
+              "Last updated :",
+              strftime(Sys.time(), format = "%Y-%m-%d %H:%M:%S")
+            ),
+            style = list(
+              fontSize = 12,
+              marginTop = 4
+            )
+          )
+        ),
+        style = list(marginTop = "4", color = "#b4b4b4")
+      ),
+      htmlBr(),
+      htmlDiv(
+        list(
+          htmlTable(
+            rows
+          )
+        ),
+        style = list(
+          height = "400px",
+          overflowY = "scroll",
+          border = "1px solid #56585d"
+        )
+      )
+    ),
+    style = list(height = "100%")
+  )
+}
+
+# Get the news from newsapi.org
+updateNews <- function(){
+  apiURL <- "https://newsapi.org/v2/top-headlines?sources=bbc-news&apiKey=da8e2e705b914f9f86ed2e9692e66012" 
+  r <- GET(apiURL)
+  jsonData <- content(r, "parsed")$articles
+  # include tryCatch for when the api breaks
+  out <- tryCatch({
+    d <- data.frame(do.call("rbind", jsonData))
+    d <- d[, c("title", "url")]
+  },
+    error = function(x){
+      d <- data.frame(title = "news api not available",
+                      url = apiURL)
+    }
+  )
+  generateNewsTable(out)
+}
+
+# Header for left div (table header with live clock)
+getHeader <- function(t = Sys.time()){
+  htmlDiv(
+    list(
+      htmlP(
+        strftime(Sys.time(), format = "%H:%M:%S"),
+        id = "live_clock",
+        className = "four columns",
+        style = list(color = "#b4b4b4", textAlign = "center")
+      ),
+      htmlP(
+        "Bid",
+        className = "four columns",
+        style = list(color = "white", textAlign = "center")
+      ),
+      htmlP(
+        "Ask",
+        className = "four columns",
+        style = list(color = "white", textAlign = "center")
+      )
+    )
+  )
+}
+
+# color of Bid & Ask rates (depending on previous bids and asks)
+getColor <- function(a, b){
+  ifelse(
+    a == b,
+    "white",
+    ifelse(
+      a > b,
+      "#45df7e",
+      "#da5657"
+    )
+  )
+}
+
+# Row or previous 10 rows of currency pair dataset
+getAskBid <- function(currency_pair, index, modal = FALSE){
+  if (modal == FALSE){
+    currency_pair[index, ]
+  } else {
+    currency_pair[c( (index - 10):index), ]
```

Comment:
```suggestion
    currency_pair[c( (index - 100):index), ]
```
Displaying last 100 ask/bid pairs instead of 10, makes the order graph more insightful

Reply:
That's probably true, I'm just wondering whether we should change the functionality relative to the python app, if people are going to use this code to try to reproduce what they see on the gallery? @rpkyle, what do you think?

Reply:
@sacul-git I'm inclined to leave as-is for now, but we could open an issue to address @CanerIrfanoglu's point, and then modify both apps, as you suggest here.